src/main/webapp/isolate/request/scripts/TicketFactory.js

```diff
@@ -0,0 +1,112 @@
+ * form.
+ * 
+ * This class __must__ be instantiated in order to function correctly.
+ */
+export class TicketFactory {
+  /**
+   * Construct a TicketFactory.
+   */
+  constructor() {
+    /* The number of tickets created. */
+    this.ticketCount = 0;
+    /* The number of times the page has attempted to fetch the ticket template
+     * from the server. */
+    this.ticketTemplateFetchAttempts = 0;
+    /* The template for request tickets. */
+    this.ticketTemplate = null;
+  }
+
+  /**
+   * Cache the Ticket Template, if necessary. If the Ticket Template already
+   * exists, calling this function is a no-op.
+   * 
+   * @throws Will throw an Error if the server cannot retrieve the
+   *     Ticket Template.
+   */
+  async cacheTicketTemplate() {
+    /* cacheTicketTemplate is recursive. If ten or more failed attempts are made
+     * to fetch the template, the process is aborted. */
+    if (this.ticketTemplateFetchAttempts >= 10) {
+      throw (Error('Ticket template cannot be fetched from the server.'));
+    }
+
+    /* If the ticketTemplate is null or undefined, attempt to fetch it from the
+     * server. */
+    if (this.ticketTemplate === null ||
+        typeof this.ticketTemplate === 'undefined') {
+      const response = await fetch('request-ticket.html');
+
+      /* Each time the server is queried, increment the number of fetch attempts
+       */
+      this.ticketTemplateFetchAttempts++;
+
+      /* If the request fails, try again. */
+      if (!response.ok) {
+        this.cacheTicketTemplate();
+      }
+      /* If the request succeeds, attempt to parse the body of the response */
+      else {
+        const text = await response.text();
+
+        /* If parsing the response body fails, try again. */
+        if (text === null || typeof text === 'undefined') {
+          this.cacheTicketTemplate();
+        }
+        /* If parsing the response body succeeds, cache the response body. */
+        else {
+          this.ticketTemplate = text;
+        }
+      }
+    }
+  }
+
+  /**
+   * Creates a Ticket.
+   * @return a new Ticket
+   * @throws Will throw an Error if the server cannot retrieve the
+   *     Ticket Template.
+   */
+  async createTicket() {
+    try {
+      await this.cacheTicketTemplate();
+    } catch (error) {
+      throw (Error(`Ticket creation failed with error\n:${error}`));
+    }
+
+    const ticket = document.createElement('fieldset');
+    ticket.innerHTML = this.ticketTemplate;
+
+    /* Rename each ticket's input element */
+    for (let child of ticket.children) {
+      /*
+       * The name of the tickets always resembles a 2D array. This ensures that
+       * the browser can format the request properly, such that all of the
+       * tickets are recieved on the backend within a single POST request.
+       */
+      child.name = `tickets[${this.ticketCount}][${child.name}]`;
+    }
+
+    this.ticketCount++;
+    return ticket;
```

Comment:
What is the purpose of this code?

Reply:
It creates a ticket for insertion into the request form. The renaming of the tickets is necessary to ensure that, on posting, the elements of each ticket (title, description etc.) are grouped on a per-ticket basis.